### PR TITLE
Update serde.py for Secure Aggregation

### PR DESCRIPTION
**Pull request by panh99**
### Description
No serialization/deserialization functions for Secure Aggregation field `sa` in `Task` message.

### Related issues/PRs
Add SecureAggregation field to Task message #1893
adopt TaskIns/TaskRes-based communications on the client side #2007

Separate `validator.py` update: Update validator for SA #2151

### Explanation
Add `Value` in `typing.py`

Add functions in `serde.py` for `Value` message

Add unit tests for `Value` message and `Dict[str, Value]` in `serde_test.py`

### Checklist

- [x] Implement proposed change
- [x] Write tests
- [ ] Update [documentation](https://flower.dev/docs/writing-documentation.html)
- [ ] Update [changelog](https://github.com/adap/flower/blob/main/doc/source/changelog.rst)
- [x] Make CI checks pass
- [x] Ping maintainers on [Slack](https://flower.dev/join-slack/) (channel `#contributions`)

### CODE REVIEW

**Review by tanertopal — COMMENTED** [review]
@panh99  I am less familiar with this part of Flower then @danieljanes  is. It looks good, but I'd wait for his review before merging.

File: src/py/flwr/server/utils/validator.py [hunk 1]
```diff
@@ -21,6 +21,7 @@
 
 
 # pylint: disable-next=too-many-branches
+# pylint: disable-next=too-many-statements
```
Comment (danieljanes) [inline src/py/flwr/server/utils/validator.py:24]: This should work too:
```suggestion
# pylint: disable-next=too-many-branches,too-many-statements
```

File: src/py/flwr/server/utils/validator.py [hunk 2]
```diff
@@ -63,12 +64,21 @@ def validate_task_ins_or_res(tasks_ins_res: Union[TaskIns, TaskRes]) -> List[str
...
-        if not tasks_ins_res.task.HasField("legacy_server_message"):
+        # Content check
+        has_fields = {
+            "sa": tasks_ins_res.task.HasField("sa"),
+            "legacy_server_message": tasks_ins_res.task.HasField(
+                "legacy_server_message"
+            ),
+        }
+        if sum(has_fields.values()) == 0:
+            err_msg = ", ".join([f"`{o}`" for o in has_fields])
```
Comment (danieljanes) [inline src/py/flwr/server/utils/validator.py:75]:
```suggestion
            err_msg = ", ".join([f"`{field}`" for field in has_fields])
```

File: src/py/flwr/server/utils/validator.py [hunk 3]
```diff
@@ -63,12 +64,21 @@ def validate_task_ins_or_res(tasks_ins_res: Union[TaskIns, TaskRes]) -> List[str
...
-        # Legacy ServerMessage
-        if not tasks_ins_res.task.HasField("legacy_server_message"):
+        # Content check
+        has_fields = {
+            "sa": tasks_ins_res.task.HasField("sa"),
+            "legacy_server_message": tasks_ins_res.task.HasField(
+                "legacy_server_message"
+            ),
+        }
+        if sum(has_fields.values()) == 0:
```
Comment (danieljanes) [inline src/py/flwr/server/utils/validator.py:74]: Let's make this more boring:
```suggestion
        if not (has_fields["sa"] or has_fields["legacy_server_message"]):
```

**Review by danieljanes — CHANGES_REQUESTED** [review]
Let's move the changes in `flwr.common` into a new PR and focus this one to first merge the changes in `validator.py`. In addition to the review comments, it would be good to add unit tests to cover the changes.

File: src/py/flwr/common/serde.py [hunk 4]
```diff
@@ -483,3 +484,87 @@ def scalar_from_proto(scalar_msg: Scalar) -> typing.Scalar:
...
+__python_list_type_to_message_and_field_name = {
+    float: (Value.DoubleList, "double_list"),
+    int: (Value.Sint64List, "sint64_list"),
+    bool: (Value.BoolList, "bool_list"),
+    str: (Value.StringList, "string_list"),
+    bytes: (Value.BytesList, "bytes_list"),
+}
+
+
+def __check_value(value: typing.Value) -> None:
```
Comment (danieljanes) [inline src/py/flwr/common/serde.py:510]:
```suggestion
def _check_value(value: typing.Value) -> None:
```

File: src/py/flwr/common/serde.py [hunk 5]
```diff
@@ -483,3 +484,87 @@ def scalar_from_proto(scalar_msg: Scalar) -> typing.Scalar:
...
+    if isinstance(value, (list, tuple)) and isinstance(
+        value[0], tuple(__python_type_to_field_name.keys())
+    ):
+        data_type = type(value[0])
+        for element in value:
+            if isinstance(element, data_type):
+                continue
+            raise Exception(
+                f"Inconsistent type: the types of elements in the list must be the same"
+                f"(expect {data_type}, but get {type(element)})"
```
Comment (danieljanes) [inline src/py/flwr/common/serde.py:522]:
```suggestion
                f"(expected {data_type}, but got {type(element)})"
```

**Review by danieljanes — CHANGES_REQUESTED** [review]

**Review by danieljanes — APPROVED** [review]

**Pull request merged**